src/events.rs

```diff
@@ -60,30 +59,24 @@ impl Events {
 /// created events emitted by the [`Context`] will only be delivered to one of the
 /// `EventEmitter`s.
 ///
-/// The `EventEmitter` is also a [`Stream`], so a typical usage is in a `while let` loop.
-///
 /// [`Context`]: crate::context::Context
 /// [`Context::get_event_emitter`]: crate::context::Context::get_event_emitter
-/// [`Stream`]: futures::stream::Stream
-#[derive(Debug, Clone)]
-#[pin_project]
-pub struct EventEmitter(#[pin] Receiver<Event>);
+#[derive(Debug)]
+pub struct EventEmitter(Mutex<async_broadcast::Receiver<Event>>);
 
 impl EventEmitter {
     /// Async recv of an event. Return `None` if the `Sender` has been dropped.
     pub async fn recv(&self) -> Option<Event> {
-        self.0.recv().await.ok()
+        let mut lock = self.0.lock().await;
+        lock.recv().await.ok()
     }
-}
 
-impl futures::stream::Stream for EventEmitter {
-    type Item = Event;
-
-    fn poll_next(
-        self: std::pin::Pin<&mut Self>,
-        cx: &mut std::task::Context<'_>,
-    ) -> std::task::Poll<Option<Self::Item>> {
-        self.project().0.poll_next(cx)
+    /// Tries to receive an event without blocking.
+    ///
+    /// Returns `None` if no events are available for reception.
+    pub async fn try_recv(&self) -> Option<Event> {
+        let mut lock = self.0.lock().await;
```

Comment:
So, `EventEmitter::recv()` can be called in parallel by multiple tasks, `try_recv()` as well, but together they can't be called in parallel as `try_recv()` would wait for `recv()`. How critical it is? Afaiu not at all as each task should use its own emitter?

Reply:
> So it is not MPMC in the same sense as ordinary channel is, multiple "consumers" cannot read from the same receiver at the same time. Maybe this should be filed as a bug to async-broadcast?

I think they meant that multiple consumers should use multiple receivers. But then yes, it's not MPMC because those receivers don't "steal" messages from each other, but receive message copies.

> This is actually a bug, at least needs a comment above `recv` that it may lock out `try_recv` if there is no easy way to solve it.

Seems that adding some flag like "is_empty: RwLock<bool>" (set by `recv()`) solves this, but then `try_recv()` can miss events if the flag isn't yet reset, but new events have already arrived.

Reply:
Even `bool` isn't needed, just `RwLock<()>`. `recv()` should do `try_recv()` first which takes a read lock and if `try_recv()` didn't succeed, take a write lock